src/aclic.adoc

```diff
@@ -937,271 +936,101 @@ The summarized behavior is the following:
-If such an interrupt exists, the hart may abort executing the tpopxret instruction and trigger the following actions:
-  - Update the `xcause.Exception Code` to the corresponding interrupt id.
-  - Perform the trap handler fetch.
-  - Start executing the trap handler, while skipping the portions of tpush that have not been reverted by the aborted tpopxret.
-
-Explicitly, {xepc}, {xpistatus} remain unchanged.
-
-NOTE: In the typical case of tail chaining, the check for another pending enabled interrupt is done after disabling interrupts at the start of the tpopxret instruction. In this case, only the conditional interrupt enablement of the tpush instruction would need to executed.
-
-Once the instruction retires, or the interrupts are disabled, no late preemption can occur.
+  * Conditionally performs a stack pointer swap by executing cspspush sp
+  * Save registers x10-x15 (these are the caller saved registers available in RVC)
+  * Adjusts the stack pointer to create the stack frame
 
-=== Stack layout
+NOTE: This extension does not require a normative stack layout.
+Interrupt handlers do not require access to the interrupted context.
 
-To ease debugging, a stack layout is recommended for the saved context.
+=== Returns from interrupt handlers
 
-NOTE: This stack layout is not mandated to allow e.g. shadow register sets
+When executing an {ipopxret} instruction the context save during the trap is restored and a return is executed.
 
-[source]
-----
-* Smnip/Ssnip not present:
+NOTE: An explicit interrupt return instruction is included, rather than modifying the xret behavior,
+so that the hart does not need to do bookkeeping whether it is in an interrupt or exception context.
 
-** XLEN = 32:
+It performs the following actions in order:
 
-stack pointer adjustment: 32
+  * Restores registers x10-x15
+  * Adjust the stack pointer to destroy the stack frame
+  * Conditionally performs a stack pointer swap by executing cspspop sp
+  * Executes an xret according to the current level
 
-24: x15
-20: x14
-16: x13
-12: x12
-8: x11
-4: x10
+NOTE: As this instruction is intended to be executed at the end of the handler, interrupts are typically disabled during its execution.
+It is however an implementation choice to perform tail-chaining to the next highest priority pending-enabled interrupt upon this instruction.
 
-** XLEN = 64:
+=== C interrupt handlers
 
-stack pointer adjustment: 48
+When this extension is implemented, a compiler may assume for any interrupt handler defined with the interrupt attribute like
 
-48: x15
-40: x14
-32: x13
-24: x12
-16: x11
-8: x10
+[source,c]
+----
+void interrupt_handler() __attribute__((interrupt));
+----
 
-* Smnip/Ssnip present:
+that the registers x10-x15 have been saved.
+Additionally, the compiler does not need to restore these registers.
+Finally, it should generate an {ixpopret} instruction at the end of the function instead of an {xret}.
 

```

Comment:
That's going to get messy quickly. Not to mention that implementing trap handler in C will be impossible.


Better to do:
```
void interrupt_handler() __attribute__((interrupt("CLIC-something")));
```
or
```
void interrupt_handler() __attribute__((interrupt("CLIC-something"), prestacked("x10-x15")));
```